File: creatoriq_backend/app/schemas/notification.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# ============================================================
# 1. Create Notification
# ============================================================

class NotificationCreate(BaseModel):
    user_id: int
    contract_id: int | None = None
    obligation_id: int | None = None
    notification_type: str
    title: str
    message: str
    scheduled_at: datetime | None = None


# ============================================================
# 2. Notification Response
# ============================================================

class NotificationResponse(BaseModel):
    id: int
    user_id: int
    contract_id: int | None
    obligation_id: int | None
    notification_type: str
    title: str
    message: str
    status: str
    scheduled_at: datetime | None
    sent_at: datetime | None
    read_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ============================================================
# 3. Notification List Response
# ============================================================

class NotificationListResponse(BaseModel):
    id: int
    user_id: int
    contract_id: int | None
    obligation_id: int | None
    notification_type: str
    title: str
    message: str
    status: str
    scheduled_at: datetime | None
    sent_at: datetime | None
    read_at: datetime | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/contractiq-frontend/src/app/services/notification.ts
import { Injectable } from '@angular/core';
import {
  HttpClient
} from '@angular/common/http';

import {
  Observable
} from 'rxjs';


// =========================
// NOTIFICATION INTERFACE
// =========================

export interface Notification {

  id: number;

  user_id: number;

  contract_id: number | null;

  obligation_id: number | null;

  notification_type: string;

  title: string;

  message: string;

  status: string;

  scheduled_at: string | null;

  sent_at: string | null;

  read_at: string | null;

  created_at: string;

  updated_at: string;

}


// =========================
// MARK ALL RESPONSE
// =========================

export interface MarkAllReadResponse {

  message: string;

  count: number;

}


// =========================
// SERVICE
// =========================

@Injectable({
  providedIn: 'root'
})


export class NotificationService {

  private apiUrl =
    'http://127.0.0.1:8000';


  constructor(
    private http: HttpClient
  ) {}


  // =========================
  // GET ALL
  // =========================

  getNotifications():
    Observable<Notification[]> {

    return this.http.get<Notification[]>(
      `${this.apiUrl}/notifications`
    );

  }


  // =========================
  // GET SINGLE
  // =========================

  getNotification(
    notificationId: number
  ): Observable<Notification> {

    return this.http.get<Notification>(
      `${this.apiUrl}/notifications/${notificationId}`
    );

  }


  // =========================
  // MARK ONE AS READ
  // =========================

  markAsRead(
    notificationId: number
  ): Observable<Notification> {

    return this.http.patch<Notification>(
      `${this.apiUrl}/notifications/${notificationId}/read`,
      {}
    );

  }


  // =========================
  // MARK ALL AS READ
  // =========================

  markAllAsRead():
    Observable<MarkAllReadResponse> {

    return this.http.patch<MarkAllReadResponse>(
      `${this.apiUrl}/notifications/read-all`,
      {}
    );

  }

}